core/object/thread_runner.cpp

```diff
@@ -0,0 +1,207 @@
+/* distribute, sublicense, and/or sell copies of the Software, and to    */
+/* permit persons to whom the Software is furnished to do so, subject to */
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "thread_runner.h"
+
+#include "core/os/os.h"
+
+ThreadRunner *ThreadRunner::singleton = nullptr;
+
+void ThreadRunner::_process_task() {
+	task_mutex.lock();
+	Task *task = tasks_waiting.first()->self();
+	tasks_waiting.remove(tasks_waiting.first());
+	task_mutex.unlock();
+
+	if (task->native_func) {
+		task->native_func(task->native_func_userdata);
+	} else {
+		Callable::CallError ce;
+		Variant ret;
+		task->callable.call(nullptr, 0, ret, ce);
+	}
+
+	task->completed = true;
+	task->done_semaphore.post();
+}
+
+void ThreadRunner::_thread_function(void *p_user) {
+	ThreadData *thread = static_cast<ThreadData *>(p_user);
+	while (true) {
+		thread->scheduler->task_available_semaphore.wait();
+		if (thread->scheduler->exit_threads.load()) {
+			break;
+		}
+		thread->scheduler->_process_task();
+	}
+}
+
+ThreadRunner::TaskID ThreadRunner::add_native_task(void (*p_func)(void *), void *p_userdata, const String &p_description) {
+	task_mutex.lock();
+	// Get a free task
+	Task *task = task_allocator.alloc();
+	TaskID id = last_task++;
+	task->native_func = p_func;
+	task->native_func_userdata = p_userdata;
+	task->description = p_description;
+	tasks.insert(id, task);
+	tasks_waiting.add_last(&task->task_elem);
+	task_mutex.unlock();
+
+	task_available_semaphore.post();
+	return id;
+}
+
+ThreadRunner::TaskID ThreadRunner::add_task(const Callable &p_action, const String &p_description) {
+	task_mutex.lock();
+	// Get a free task
+	Task *task = task_allocator.alloc();
+	TaskID id = last_task++;
+	task->callable = p_action;
+	task->description = p_description;
+	tasks.insert(id, task);
+	tasks_waiting.add_last(&task->task_elem);
+	task_mutex.unlock();
+
+	task_available_semaphore.post();
+	return id;
+}
+
+bool ThreadRunner::is_task_completed(TaskID p_task_id) const {
```

Comment:
I think it should be fine, since the actual one that does polling actually contains wait in the name. This means you could deduce that does not block relatively easily.

Reply:
Misunderstanding.  I meant the API suggests that client code that wants to poll for status use this function.  As such, they might call this a lot of time, so give them a way to at least not lock for *every* task they are waiting on.  Allow checking all tasks you are waiting for in one go.